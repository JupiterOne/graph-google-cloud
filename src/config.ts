import {
  IntegrationExecutionContext,
  IntegrationLogger,
} from '@jupiterone/integration-sdk-core';
import { ServiceUsageClient } from './steps/service-usage/client';
import { IntegrationConfig, SerializedIntegrationConfig } from './types';
import { deserializeIntegrationConfig } from './utils/integrationConfig';
import { google } from 'googleapis';
import { handleApiClientError } from './google-cloud/client';
import { ServiceUsageListFilter } from './google-cloud/types';

export async function executeTestRequest(
  config: IntegrationConfig,
  logger: IntegrationLogger,
) {
  try {
    const client = google.serviceusage({ version: 'v1', retry: false });
    const googleClient = new ServiceUsageClient({ config }, logger);
    const auth = await googleClient.getAuthenticatedServiceClient();

    await client.services.list({
      parent: `projects/${config.serviceAccountKeyConfig.project_id}`,
      pageSize: 200,
      auth: auth,
      filter: ServiceUsageListFilter.ENABLED,
    });
  } catch (err) {
    throw handleApiClientError(err);
  }
}

export async function validateInvocation(
  context: IntegrationExecutionContext<SerializedIntegrationConfig>,
) {
  const { instance, logger } = context;
  const { config: serializedIntegrationConfig } = instance;
  const config = (context.instance.config = deserializeIntegrationConfig(
    serializedIntegrationConfig,
  ));

  await executeTestRequest(config, logger);
}
