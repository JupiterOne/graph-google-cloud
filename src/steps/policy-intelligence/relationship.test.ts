import {
  createMockStepExecutionContext,
  executeStepWithDependencies,
  filterGraphObjects,
  Recording,
  StepTestConfig,
} from '@jupiterone/integration-sdk-testing';
import { integrationConfig } from '../../../test/config';
import {
  getMatchRequestsBy,
  setupGoogleCloudRecording,
} from '../../../test/recording';
import {
  POLICY_INTELLIGENCE_ANALYZER_CLASS,
  POLICY_INTELLIGENCE_ANALYZER_TYPE,
  RELATIONSHIP_TYPE_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER,
  STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER,
  STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
  STEP_POLICY_INTELLIGENCE_ANALYZER_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
} from './constants';
import { IntegrationConfig, invocationConfig } from '../..';
import {
  ExplicitRelationship,
  Relationship,
} from '@jupiterone/integration-sdk-core';
import {
  buildProjectHasPolicyAnalyzerRelationship,
  fetchPolicyAnalyzer,
} from '.';
import {
  fetchResourceManagerProject,
  PROJECT_ENTITY_TYPE,
} from '../resource-manager';

describe(`#${STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER}`, () => {
  let recording: Recording;
  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER,
    });
  });
  afterEach(async () => {
    await recording.stop();
  });
  function separateRelationships(collectedRelationships: Relationship[]) {
    const { targets: directRelationships } = filterGraphObjects(
      collectedRelationships,
      (r) => !r._mapping,
    ) as {
      targets: ExplicitRelationship[];
    };
    return {
      directRelationships,
    };
  }
  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: {
        ...integrationConfig,
        serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
          'j1-gc-integration-dev-v2',
          'j1-gc-integration-dev-v3',
        ),
        serviceAccountKeyConfig: {
          ...integrationConfig.serviceAccountKeyConfig,
          project_id: 'j1-gc-integration-dev-v3',
        },
      },
    });
    await fetchPolicyAnalyzer(context);
    await fetchResourceManagerProject(context);
    await buildProjectHasPolicyAnalyzerRelationship(context);
    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();
    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === PROJECT_ENTITY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_project' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          projectId: { type: 'string' },
          name: { type: 'string' },
          displayName: { type: 'string' },
          parent: { type: 'string' },
          lifecycleState: { type: 'string' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });
    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === POLICY_INTELLIGENCE_ANALYZER_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: POLICY_INTELLIGENCE_ANALYZER_CLASS,
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: POLICY_INTELLIGENCE_ANALYZER_TYPE },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          summary: { type: 'string' },
          category: { type: 'array' },
          function: { type: 'array' },
          projectId: { type: 'string' },
          activityType: { type: 'string' },
          observationPeriod: { type: 'string' },
        },
      },
    });
    const { directRelationships } = separateRelationships(
      context.jobState.collectedRelationships,
    );
    expect(
      directRelationships.filter(
        (e) =>
          e._type ===
          RELATIONSHIP_TYPE_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const:
              RELATIONSHIP_TYPE_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER,
          },
        },
      },
    });
  });
});

describe(`Policy-intelligence#${STEP_POLICY_INTELLIGENCE_ANALYZER_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY}`, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });
  test(STEP_POLICY_INTELLIGENCE_ANALYZER_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_POLICY_INTELLIGENCE_ANALYZER_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfig),
        recordFailedRequests: true,
      },
    });
    const stepTestConfig: StepTestConfig = {
      stepId: STEP_POLICY_INTELLIGENCE_ANALYZER_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
      instanceConfig: integrationConfig,
      invocationConfig: invocationConfig as any,
    };
    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  }, 300000);
});


describe(`Policy-intelligence#${STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY}`, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });
  test(STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfig),
        recordFailedRequests: true,
      },
    });
    const stepTestConfig: StepTestConfig = {
      stepId: STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
      instanceConfig: integrationConfig,
      invocationConfig: invocationConfig as any,
    };
    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  }, 300000);
});


// describe(
//   `#${STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY}`,
//   () => {
//     let recording: Recording;
//     beforeEach(() => {
//       recording = setupGoogleCloudRecording({
//         directory: __dirname,
//         name: STEP_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
//         options: {
//           matchRequestsBy: getMatchRequestsBy(integrationConfig),
//           recordFailedRequests: true,
//         },
//       });
//     });
//     afterEach(async () => {
//       await recording.stop();
//     });
//     function separateRelationships(collectedRelationships: Relationship[]) {
//       const { targets: directRelationships } = filterGraphObjects(
//         collectedRelationships,
//         (r) => !r._mapping,
//       ) as {
//         targets: ExplicitRelationship[];
//       };
//       return {
//         directRelationships,
//       };
//     }
//     test('should collect data', async () => {
//       const context = createMockStepExecutionContext<IntegrationConfig>({
//         instanceConfig: {
//           ...integrationConfig,
//           serviceAccountKeyFile:
//             integrationConfig.serviceAccountKeyFile.replace(
//               'j1-gc-integration-dev-v2',
//               'j1-gc-integration-dev-v3',
//             ),
//           serviceAccountKeyConfig: {
//             ...integrationConfig.serviceAccountKeyConfig,
//             project_id: 'j1-gc-integration-dev-v3',
//           },
//         },
//       });
//       await fetchPolicyAnalyzerActivity(context);
//       await fetchResourceManagerProject(context);
//       await buildProjectHasPolicyAnalyzerActivityRelationship(context);
//       expect({
//         numCollectedEntities: context.jobState.collectedEntities.length,
//         numCollectedRelationships:
//           context.jobState.collectedRelationships.length,
//         collectedEntities: context.jobState.collectedEntities,
//         collectedRelationships: context.jobState.collectedRelationships,
//         encounteredTypes: context.jobState.encounteredTypes,
//       }).toMatchSnapshot();
//       expect(
//         context.jobState.collectedEntities.filter(
//           (e) => e._type === PROJECT_ENTITY_TYPE,
//         ),
//       ).toMatchGraphObjectSchema({
//         _class: ['Account'],
//         schema: {
//           additionalProperties: false,
//           properties: {
//             _type: { const: 'google_cloud_project' },
//             _rawData: {
//               type: 'array',
//               items: { type: 'object' },
//             },
//             projectId: { type: 'string' },
//             name: { type: 'string' },
//             displayName: { type: 'string' },
//             parent: { type: 'string' },
//             lifecycleState: { type: 'string' },
//             createdOn: { type: 'number' },
//             updatedOn: { type: 'number' },
//           },
//         },
//       });
//       expect(
//         context.jobState.collectedEntities.filter(
//           (e) => e._type === POLICY_INTELLIGENCE_ANALYZER_ACTIVITY_TYPE,
//         ),
//       ).toMatchGraphObjectSchema({
//         _class: POLICY_INTELLIGENCE_ANALYZER_ACTIVITY_CLASS,
//         schema: {
//           additionalProperties: false,
//           properties: {
//             _type: { const: POLICY_INTELLIGENCE_ANALYZER_ACTIVITY_TYPE },
//             _rawData: {
//               type: 'array',
//               items: { type: 'object' },
//             },
//             name: { type: 'string' },
//             summary: { type: 'string' },
//             category: { type: 'string' },
//             internal: { type: 'boolean' },
//             serviceAccountId: { type: 'string' },
//             endTime: { type: 'string' },
//             startTime: { type: 'string' },
//           },
//         },
//       });
//       const { directRelationships } = separateRelationships(
//         context.jobState.collectedRelationships,
//       );
//       expect(
//         directRelationships.filter(
//           (e) =>
//             e._type ===
//             RELATIONSHIP_TYPE_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
//         ),
//       ).toMatchDirectRelationshipSchema({
//         schema: {
//           properties: {
//             _class: { const: 'HAS' },
//             _type: {
//               const:
//                 RELATIONSHIP_TYPE_GOOGLE_CLOUD_PROJECT_HAS_POLICY_INTELLIGENCE_ANALYZER_ACTIVITY,
//             },
//           },
//         },
//       });
//     }, 300000);
//   },
// );
