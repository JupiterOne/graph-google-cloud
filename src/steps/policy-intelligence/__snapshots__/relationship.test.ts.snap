// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetch-project-has-policy-intelligence-analyzer-relation should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "W/"8f56735f6f3d890d"",
            "labels": {
              "firebase": "enabled",
            },
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2024-05-15T11:48:40.174923Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "tag.firebase": "enabled",
      "updatedOn": 1715773720174,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "google_cloud_project",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
